02-express-mongo-crud/express-params-queries/27-28/README.md

```diff
@@ -0,0 +1,355 @@
+});
+
+app.listen(3000, function () {
+  console.log("Go to localhost:3000/");
+});
+
+```
+
+## Request (URL) Parameters
+
+What if we want to create an app that can dynamically say hello to anyone?
+
+* Using **url parameters** add a dynamic route to the application, indicated by `:` and the variable name you want to use, we'll use `:name` for the example below.
+
+```js
+app.get("/greet/:name", function (req, res) {
+  res.send( "Hello, " + req.params.name );
+});
+```
+
+Here we are seeing the first introduction to parameters that the application can identify. In the following route `:name` is considered a route parameter. We can access it using `req.params.name`.
+
+| Request Type | Request Path | Response
+| :--- | :--- | :--- |
+| `GET` | `/greet/:name` | `Hello, :name` |
+
+
+## Query Parameters
+
+Generally, you don't want to cram everything into a route. Just imagine when there are multiple parameters in route. Or maybe we don't care about getting the order of the parameters correct. Luckily, there are **query parameters** you can include with each request.
+
+Let's see query params in action. Go to <a href="https://google.com/search?q=kittens&tbm=isch" target="_blank">https://google.com/search?q=kittens&tbm=isch</a>
+
+* `?` denotes the beginning of the query parameters.
+* `=` indicates an assignment; anything to the left is the key, while the right represents the value.
+* `&` allows for the input of multiple parameters, separating each.
+
+Let's add our first route to practice query params.
+
+```js
+app.get("/thank", function (req, res) {
+  var name = req.query.name;
+  res.send("Thank you, " + name);
+});
+```
+
+Reset your server and go to <a href="localhost:3000/thank?name=jane" target="_blank">localhost:3000/thank?name=jane</a>. Note how we can now define parameters in the url after a `?`.
+
+## Choosing between request params and query params
+
+* **request params:** `http://localhost:3000/icecream/flavors/:flavor`
+
+* **query params:** `http://localhost:3000/icecream?flavor=SOMEFLAVOR`
+
+Generally if the parameter is identifying a specific _entity_ or _resource_ on it's own, you most likely want request params.  If it is an optional param for this route, you generally should use query params.
+
+Common cases for **Request Params**:
+* **database/item IDs**            -  GET `/contacts/348`
+* **major components of the app**  -  GET `/posts/33`
+  * resources!
+* **expressing hierarchies**       -  GET `/departments/44/employees/2`
+
+> Request Params are not always IDs either; names can be used in some systems: `/departments/accounting/employees/LiuJanice`
+
+Common cases for **Query Params**:
+* **searches**             -  GET `/?q=kittens`
+* **optional selections**  -  GET `/calculator?lang=en`
+* **pagination**           -  GET `/posts?start=1&size=10`
+* **other limits**         -  GET `/posts?since=2015-11-29`
+
+Finally you might use them both together in some cases: `/posts/33/comments?limit=50`
+
+## Middleware
```

Comment:
So in 25 we did middleware here very briefly. However if we're doing body-parser we have to first talk about POST and that really would need it's own section. 27 hasn't seen POST used yet. @justincastilla should we just take out body-parser?
 Add a line note